### PR TITLE
change getprop interface

### PR DESCRIPTION
**Pull request by nevermore3**
use smart pointer manage` vector<vertexProp>` and `vector<edgeProp>`
```
return EMPTY.                                 when  smart pointer is nullptr

[ FORBIT ]return ALL props in schema.         when  the content pointed to by the smart pointer is empty 

return SPECIFIC attributes             when  the content pointed to by the smart pointer  has wanted attributes
```

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
> use smart pointer manage` vector<vertexProp>` and `vector<edgeProp>`
> 
> ```
> return EMPTY.                                 when  smart pointer is nullptr
> 
> [ FORBIT ]return ALL props in schema.         when  the content pointed to by the smart pointer is empty 
> 
> return SPECIFIC attributes             when  the content pointed to by the smart pointer  has wanted attributes
> ```

I think return all properties when the pointer is null, and return empty properties when the container is empty is better.

**Comment by nevermore3** [comment]
> > use smart pointer manage` vector<vertexProp>` and `vector<edgeProp>`
> > ```
> > return EMPTY.                                 when  smart pointer is nullptr
> > 
> > [ FORBIT ]return ALL props in schema.         when  the content pointed to by the smart pointer is empty 
> > 
> > return SPECIFIC attributes             when  the content pointed to by the smart pointer  has wanted attributes
> > ```
> 
> I think return all properties when the pointer is null, and return empty properties when the container is empty is better.


These are defined by thrift

**Comment by Shylock-Hg** [comment]
> > > use smart pointer manage` vector<vertexProp>` and `vector<edgeProp>`
> > > ```
> > > return EMPTY.                                 when  smart pointer is nullptr
> > > 
> > > [ FORBIT ]return ALL props in schema.         when  the content pointed to by the smart pointer is empty 
> > > 
> > > return SPECIFIC attributes             when  the content pointed to by the smart pointer  has wanted attributes
> > > ```
> > 
> > 
> > I think return all properties when the pointer is null, and return empty properties when the container is empty is better.
> 
> These are defined by thrift

These could modified to better.

File: src/planner/plan/Query.cpp [hunk 1]
```diff
@@ -115,21 +114,17 @@ void GetVertices::cloneMembers(const GetVertices& gv) {
...
     src_ = qctx_->objPool()->add(gv.src()->clone().release());
 
-    std::vector<storage::cpp2::VertexProp> props;
-    const auto& gvProps = gv.props();
-    props.reserve(gvProps.size());
-    for (const auto& prop : gvProps) {
-        props.emplace_back(prop);
+    if (gv.props_) {
+        auto vertexProps = *gv.props_;
+        auto vertexPropsPtr = std::make_unique<decltype(vertexProps)>(vertexProps);
```
Comment (laura-ding) [inline src/planner/plan/Query.cpp:119]: std::move(vertexProps)

File: src/planner/plan/Query.cpp [hunk 2]
```diff
@@ -115,21 +114,17 @@ void GetVertices::cloneMembers(const GetVertices& gv) {
...
-    props_ = std::move(props);
 
-    std::vector<storage::cpp2::Expr> exprs;
-    const auto& gvExprs = gv.exprs();
-    exprs.reserve(gvExprs.size());
-    for (const auto& expr : gvExprs) {
-        exprs.emplace_back(expr);
+    if (gv.exprs_) {
+        auto exprs = *gv.exprs_;
+        auto exprsPtr = std::make_unique<decltype(exprs)>(exprs);
```
Comment (laura-ding) [inline src/planner/plan/Query.cpp:125]: ditto

**Review by CPWstatic — APPROVED** [review]

File: src/planner/plan/Query.cpp [hunk 3]
```diff
@@ -158,21 +153,17 @@ void GetEdges::cloneMembers(const GetEdges& ge) {
...
     dst_ = qctx_->objPool()->add(ge.dst()->clone().release());
 
-    std::vector<storage::cpp2::EdgeProp> props;
-    const auto& geProps = ge.props();
-    props.reserve(geProps.size());
-    for (const auto& prop : geProps) {
-        props.emplace_back(prop);
+    if (ge.props_) {
+        auto edgeProps = *ge.props_;
+        auto edgePropsPtr = std::make_unique<decltype(edgeProps)>(edgeProps);
```
Comment (laura-ding) [inline src/planner/plan/Query.cpp:158]: ditto

File: src/planner/plan/Query.h [hunk 4]
```diff
@@ -18,6 +18,10 @@
  * All query-related nodes would be put in this file,
  * and they are derived from PlanNode.
  */
+using VertexPropsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::VertexProp>>;
+using EdgePropsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::EdgeProp>>;
+using StatPropsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::StatProp>>;
+using ExprsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::Expr>>;
```
Comment (yixinglu) [inline src/planner/plan/Query.h:24]: declare these types inside namespace

File: src/util/SchemaUtil.h [hunk 5]
```diff
@@ -16,6 +16,8 @@
 #include "common/meta/NebulaSchemaProvider.h"
 #include "parser/MaintainSentences.h"
 
+using VertexPropsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::VertexProp>>;
+using EdgePropsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::EdgeProp>>;
```
Comment (yixinglu) [inline src/util/SchemaUtil.h:20]: duplicate declaration within Query.h

File: src/validator/FetchEdgesValidator.h [hunk 6]
```diff
@@ -11,6 +11,8 @@
 #include "parser/TraverseSentences.h"
 #include "validator/Validator.h"
 
+using EdgePropsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::EdgeProp>>;
+using ExprsPtr = std::unique_ptr<std::vector<nebula::storage::cpp2::Expr>>;
```
Comment (yixinglu) [inline src/validator/FetchEdgesValidator.h:15]: ditto

**Review by laura-ding — DISMISSED** [review]

File: src/planner/plan/Query.cpp (hunk 1 above)
Comment (nevermore3) [reply src/planner/plan/Query.cpp:119]: done

File: src/planner/plan/Query.h (hunk 4 above)
Comment (nevermore3) [reply src/planner/plan/Query.h:24]: done

File: src/util/SchemaUtil.h (hunk 5 above)
Comment (nevermore3) [reply src/util/SchemaUtil.h]: done

**Review by laura-ding — DISMISSED** [review]

**Review by CPWstatic — DISMISSED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by laura-ding — APPROVED** [review]

**Pull request merged**